### PR TITLE
ISPN-13373 Allow Cache availability to be configured via the console

### PR DESCRIPTION
**Pull request by dipguptaredhat**
https://issues.redhat.com/browse/ISPN-13373

* Feature to make cache available from Degraded mode

### CODE REVIEW

File: src/app/Caches/SetAvailableCache.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  isModalOpen: boolean;
+  closeModal: (boolean) => void;
+}) => {
+  const { addAlert } = useApiAlert();
+  const { reloadCaches } = useCaches();
+
+  const { t } = useTranslation();
+  const brandname = t('brandname.brandname');
+
+  const clearSetAvailableCacheModal = (ignoreDone: boolean) => {
```
Comment (karesti) [inline src/app/Caches/SetAvailableCache.tsx:23]: ignoreDone -> setAvailable done

File: src/app/Caches/SetAvailableCache.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  };
+
+  return (
+    <Modal
+      titleIconVariant={CheckCircleIcon}
+      className="pf-m-redhat-font"
+      width={'50%'}
+      isOpen={props.isModalOpen}
+      title="Make the cache Available"
```
Comment (karesti) [inline src/app/Caches/SetAvailableCache.tsx:45]: Extract these into the properties file.

Ask @oraNod for rewording please :)))

File: src/app/assets/languages/en.json [hunk 3]
```diff
@@ -61,6 +61,7 @@
     "undo-ignore" : "Show",
     "ignored-status" : "Hidden",
     "delete" : "Delete",
+    "available": "Available",
```
Comment (karesti) [inline src/app/assets/languages/en.json:64]: @oraNod which should be the action name to set it back to available when the cache is degraded ?

**Review by karesti — CHANGES_REQUESTED** [review]
we need to check the wording with @oraNod 
small commets, otherwise LGTM

File: src/app/CacheManagers/CacheTableDisplay.tsx [hunk 4]
```diff
@@ -213,7 +214,7 @@ const CacheTableDisplay = (props: {
 
     const isAdmin = ConsoleServices.security().hasConsoleACL(ConsoleACL.ADMIN, connectedUser);
     const isCreator = ConsoleServices.security().hasConsoleACL(ConsoleACL.CREATE, connectedUser);
-
+    console.log("dip",filteredCaches);
```
[the remark this answers is not in the document]
Comment (dipguptaredhat) [reply src/app/CacheManagers/CacheTableDisplay.tsx]: It's outdated code

File: src/app/Caches/SetAvailableCache.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      onClose={() => clearSetAvailableCacheModal(false)}
+      aria-label="make-cache-available"
+      description={
+        <TextContent>
+          <Text>
+            <strong>'{props.cacheName}'</strong> will be available.
```
Comment (oraNod) [inline src/app/Caches/SetAvailableCache.tsx:51]: <strong>'{props.cacheName}'</strong> will become available for read and write operations.

File: src/services/cacheService.ts [hunk 6]
```diff
@@ -578,4 +578,41 @@ export class CacheService {
...
+  */
+   public async setAvailability(cacheName: string): Promise<ActionResponse> {
+    const availabilityUrl =
+      this.endpoint +
+      '/caches/' +
+      encodeURIComponent(cacheName) +
+      '?action=set-availability&availability=AVAILABLE';
+    return this.fetchCaller.post({
+      url: availabilityUrl,
+      successMessage: `Cache ${cacheName} is available.`,
```
Comment (oraNod) [inline src/services/cacheService.ts:595]: `Cache ${cacheName} is now available.`

File: src/services/cacheService.ts [hunk 7]
```diff
@@ -578,4 +578,41 @@ export class CacheService {
...
+      errorMessage: `Unexpected error when changing availability of the cache ${cacheName}.`,
```
Comment (oraNod) [inline src/services/cacheService.ts:596]: `An error occurred while changing cache ${cacheName} availability.`

**Review by oraNod — CHANGES_REQUESTED** [review]

File: src/app/assets/languages/en.json (hunk 3 above)
Comment (oraNod) [reply src/app/assets/languages/en.json:64]: Commented in another section but "Make available" is the natural collocation imo.

File: src/app/assets/languages/en.json [hunk 8]
```diff
@@ -268,6 +269,9 @@
       "modal-enable-button" : "Enable",
       "modal-cancel-button" : "Cancel",
       "modal-description" : " Disabling rebalancing prevents the cluster from repairing data in this cache belonging to nodes leaving the cluster. It should be re-enabled as soon as possible to prevent possible data-loss in this cache."
+    },
+    "availability": {
+      "modal-available-title": "Make cache Available"
```
Comment (oraNod) [inline src/app/assets/languages/en.json:274]:
```suggestion
      "modal-available-title": "Make cache available"
```

**Review by oraNod — COMMENTED** [review]
Strings look fine to me, @dipguptaredhat Could you please externalize them to the `assets/languages/en.json` file? I think we'll be good to go then. Thanks.

File: src/app/assets/languages/en.json (hunk 3 above)
Comment (karesti) [reply src/app/assets/languages/en.json:64]: @dipguptaredhat please update the label

File: src/app/Caches/SetAvailableCache.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          onClick={handleAvailableButton}
+        >
+          Done
+        </Button>,
+        <Button
+          key="cancel"
+          variant="link"
+          onClick={() => clearSetAvailableCacheModal(false)}
+        >
+          Cancel
```
Comment (karesti) [inline src/app/Caches/SetAvailableCache.tsx:67]: labels in the properties file too please

**Review by karesti — CHANGES_REQUESTED** [review]
@dipguptaredhat a couple of small latest changes

**Comment by oraNod** [comment]
@dipguptaredhat Hi, hope you don't mind but I also pushed a commit to your branch for this one. A couple of questions, please. I've built this locally but don't see the cache availability from the console. Looking in the Cache Manager display table changes in this PR I would expect to see the availability there. Maybe I'm missing it. Could you provide a screenshot?

Also where is the "Done" button? I'm not sure where that displays or what the flow is for users to arrive to that modal. Cheers.

**Comment by dipguptaredhat** [comment]
> @dipguptaredhat Hi, hope you don't mind but I also pushed a commit to your branch for this one. A couple of questions, please. I've built this locally but don't see the cache availability from the console. Looking in the Cache Manager display table changes in this PR I would expect to see the availability there. Maybe I'm missing it. Could you provide a screenshot?
> 
> Also where is the "Done" button? I'm not sure where that displays or what the flow is for users to arrive to that modal. Cheers.



> @dipguptaredhat Hi, hope you don't mind but I also pushed a commit to your branch for this one. A couple of questions, please. I've built this locally but don't see the cache availability from the console. Looking in the Cache Manager display table changes in this PR I would expect to see the availability there. Maybe I'm missing it. Could you provide a screenshot?
> 
> Also where is the "Done" button? I'm not sure where that displays or what the flow is for users to arrive to that modal. Cheers.

For setting a cache to Available, the cache need to have this configuration
```
{
  "distributed-cache": {
    "mode": "SYNC",
    "statistics": true,
    "encoding": {
      "key": {
        "media-type": "application/x-protostream"
      },
      "value": {
        "media-type": "application/x-protostream"
      }
    },
    "partition-handling": {
      "when-split": "DENY_READ_WRITES",
      "merge-policy": "NONE"
    }
  }
}
```

Also the cache need to be in DEGRADED_MODE, 
`curl -XPOST --digest -u "admin:password" "http://localhost:11222/rest/v2/caches/Test?action=set-availability&availability=DEGRADED_MODE"`

Then the "Available" option will be shown as below
 
Screenshot from 2022-01-26 13-36-01
Screenshot from 2022-01-26 13-35-51

**Comment by oraNod** [comment]
Thanks for the screenshots, @dipguptaredhat That's actually really helpful because I was trying to get caches into degraded mode by deliberately crashing nodes (was doing that for a different PR recently). I didn't realize you could also force a cache into degraded mode via REST.

Anyway that looks good to me but I think maybe "Done" should change to "OK" because it's more of a confirmation.

**Review by oraNod — APPROVED** [review]
LGTM. Thanks @dipguptaredhat

**Review by karesti — APPROVED** [review]

**Pull request merged**